### PR TITLE
feat: support regex search

### PR DESCRIPTION
**Pull request by lievenhey**
Since we use QSortFilterProxyModel we get this for free. This patch adds support for regex search by not escaping the search term if it is prefix with %.
For the flamegraph there is a custom implementation, which changes the current QString::contains to a QRegularExpression::match call.

Closes: #666

### CODE REVIEW

**Comment by GitMensch** [comment]
That's a nice implementation! The tooltip also helps for documentation purposes (I'm still not sure if there should be a note on that in the README as well).

File: src/resultsbottomuppage.ui [hunk 1]
```diff
@@ -67,7 +67,14 @@
       <item>
        <widget class="QLineEdit" name="bottomUpSearch">
         <property name="toolTip">
-         <string>Filter the call graph tree.</string>
+         <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;Filter the call graph tree.&lt;br/&gt;Prefix with '%' to turn it into an regex.&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Comment (GitMensch) [inline src/resultsbottomuppage.ui:70]: @lievenhey you want to revert that line change here and in the resultstopdownpage.ui as the % is not used any more (and possibly add a note / updated screenshot for the README)

File: src/flamegraph.cpp [hunk 2]
```diff
@@ -546,14 +546,13 @@ struct SearchResults
     qint64 directCost = 0;
 };
 
-SearchResults applySearch(FrameGraphicsItem* item, const QString& searchValue)
+SearchResults applySearch(FrameGraphicsItem* item, const QRegularExpression& expression)
 {
     SearchResults result;
-    if (searchValue.isEmpty()) {
+    if (expression.pattern().isEmpty()) {
```
Comment (lievenhey) [inline src/flamegraph.cpp:552]: I'm not sure if this should be `!expression.isValid() || expression.pattern().isEmpty()`

File: src/resultsbottomuppage.ui (hunk 1 above)
Comment (GitMensch) [reply src/resultsbottomuppage.ui]: Still included in resultstopdownpage.ui ...

File: src/flamegraph.cpp (hunk 2 above)
Comment (GitMensch) [reply src/flamegraph.cpp:552]: if it isn't valid we should not get here, at least not when executed from the UI (the function could possibly `assert` that or do an early-exit for the invalid expression)

File: src/flamegraph.cpp [hunk 3]
```diff
@@ -805,13 +804,26 @@ FlameGraph::FlameGraph(QWidget* parent, Qt::WindowFlags flags)
             searchInput->setMinimumWidth(200);
             layout->addWidget(searchInput);
 
+            auto regexCheckBox = new QCheckBox(widget);
+            regexCheckBox->setText(tr("Regex Search"));
```
Comment (milianw) [inline src/flamegraph.cpp:808]: can you show a screenshot please - this sounds pretty wide to me - can we somehow reduce  the width that this checkbox occupies I wonder?

**Comment by milianw** [comment]
otherwise code lgtm, tests would obviously be nice to have, but I don't think we have any integration tests of that sort yet :(

File: src/flamegraph.cpp (hunk 3 above)
Comment (lievenhey) [reply src/flamegraph.cpp:808]: I mean I could shorten the text to something like `^[A-Z]$` and then add a tooltip explaining that this enables the regex search.

Comment (GitMensch) [reply src/flamegraph.cpp:808]: Even for a programmer that looks strange.
That said - the version from vscode does look quite nice (inline toggle - could still be an outside checkbox) with three short "text symbols" and the expected tooltip:

vscode search

Regex Text Symbol `.*`

Comment (milianw) [reply src/flamegraph.cpp:808]: lgtm, thanks

**Pull request merged**